impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeMetricImpl.java

```diff
@@ -0,0 +1,176 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.implcore.metrics;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.Maps;
+import io.opencensus.common.Clock;
+import io.opencensus.common.ToLongFunction;
+import io.opencensus.metrics.LabelKey;
+import io.opencensus.metrics.LabelValue;
+import io.opencensus.metrics.LongGaugeMetric;
+import io.opencensus.metrics.export.Metric;
+import io.opencensus.metrics.export.MetricDescriptor;
+import io.opencensus.metrics.export.MetricDescriptor.Type;
+import io.opencensus.metrics.export.TimeSeries;
+import io.opencensus.metrics.export.Value;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+
+/** Implementation of {@link LongGaugeMetric}. */
+public final class LongGaugeMetricImpl extends LongGaugeMetric implements Meter {
+  private final MetricDescriptor metricDescriptor;
+  private volatile Map<List<LabelValue>, Point> points = Maps.newHashMap();
+  private final int labelKeysSize;
+
+  LongGaugeMetricImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    this.labelKeysSize = labelKeys.size();
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_INT64, labelKeys);
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+
+    List<LabelValue> labelValuesCopy = new ArrayList<LabelValue>(labelValues); // Deep copy.
+    checkArgument(labelKeysSize == labelValuesCopy.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValuesCopy, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToLongFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+
+    List<LabelValue> labelValuesCopy = new ArrayList<LabelValue>(labelValues); // Deep copy.
+    checkArgument(labelKeysSize == labelValuesCopy.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValuesCopy, obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public Point getDefaultPoint() {
+    List<LabelValue> labelValues = new ArrayList<LabelValue>(labelKeysSize);
+    Point existingPoint = points.get(labelValues);
+    if (existingPoint != null) {
+      return existingPoint;
+    }
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  private synchronized <T> Point addAndGetPoint(
+      List<LabelValue> labelValues, @Nullable T obj, @Nullable ToLongFunction<T> function) {
+
+    Point existingPoint = points.get(labelValues);
+    if (existingPoint != null) {
```

Comment:
See one of my previous comment. If called twice with the same addPoint API maybe we should not return an error, if we do return and error add is not a good prefix maybe register is better.